tensorflow/core/user_ops/chol_grad.cc

```diff
@@ -0,0 +1,115 @@
+
+    //Check that input tensors are square.
+    OP_REQUIRES(context, input_tensor_l.dim_size(0) == input_tensor_l.dim_size(1), errors::InvalidArgument("Input matrix must be square."));
+    OP_REQUIRES(context, input_tensor_l_bar.dim_size(0) == input_tensor_l_bar.dim_size(1), errors::InvalidArgument("Input matrix must be square."));
+
+    //Check that input tensors are of same size.
+    OP_REQUIRES(context, input_tensor_l.dim_size(0) == input_tensor_l_bar.dim_size(0), errors::InvalidArgument("Input matrices must be of same size."));    
+
+    // Create an output tensor
+    Tensor* output_tensor = NULL;
+    
+    OP_REQUIRES_OK(context, context->allocate_output(0, input_tensor_l_bar.shape(), &output_tensor));
+
+    if (output_tensor->NumElements() == 0) {
+      // the output shape is a 0-element matrix, so there is nothing to do.
+      return;
+    }
+    
+    //The next three lines are necessary to get Eigen matrix behaviour.
+    const ConstMatrixMap input_matrix_l(input_tensor_l.flat<T>().data(), input_tensor_l.dim_size(0), input_tensor_l.dim_size(1));
+    const ConstMatrixMap input_matrix_l_bar(input_tensor_l_bar.flat<T>().data(), input_tensor_l_bar.dim_size(0), input_tensor_l_bar.dim_size(1));
+    MatrixMap output_matrix(output_tensor->template flat<T>().data(), input_tensor_l.dim_size(0), input_tensor_l.dim_size(1) );    
+
+    const lcl_size_t N = input_matrix_l.rows();
+    const lcl_size_t NB = 32;
+
+    output_matrix = input_matrix_l_bar.template triangularView<Eigen::Lower>();
+
+    
+    for ( lcl_size_t Ji = (N-NB+1) ; Ji>(1-NB); Ji-= NB )   
+    {
+        lcl_size_t J = std::max<lcl_size_t>(1, Ji);
+        lcl_size_t JB = NB - (J - Ji);
+    
+        output_matrix.block( J+JB-1, J-1, N - (J+JB-1), JB) = input_matrix_l.block( J-1, J-1, JB, JB ).adjoint().template triangularView<Eigen::Upper>().solve( output_matrix.block( J+JB-1, J-1, N - (J+JB-1), JB ).adjoint() ).adjoint();
+        output_matrix.block( J-1, J-1, JB, JB ) -= (output_matrix.block( J+JB-1, J-1, N - (J+JB-1), JB).adjoint() * input_matrix_l.block( J+JB-1, J-1, N - (J+JB-1), JB ) ).template triangularView<Eigen::Lower>();
+        output_matrix.block( (J+JB-1), 0, N - (J+JB-1), J-1 )  -=  output_matrix.block( (J+JB-1), J-1, N - (J+JB-1), JB ) * input_matrix_l.block( J-1, 0, JB, J-1 );
+        output_matrix.block( J-1, 0, JB, J-1) -= output_matrix.block( (J+JB-1), J-1, N - (J+JB-1), JB ).adjoint() * input_matrix_l.block( (J+JB-1), 0, N - (J+JB-1), J-1 ) ;
+        chol_rev_unblocked( input_matrix_l.block( J-1, J-1, JB, JB ),  output_matrix.block( J-1, J-1, JB, JB ) );
+        output_matrix.block( J-1, 0, JB, J-1 ) -= (output_matrix.block( J-1, J-1, JB, JB ) + output_matrix.block( J-1, J-1, JB, JB ).adjoint() )* input_matrix_l.block( J-1, 0, JB, J-1 );
+    }
+    
+        
+    //TODO: what to do if input_matrix_l isn't lower triangular?
```

Comment:
You can get rid of this TODO. The op should ignore any contents above the diagonal. The docstring should explicitly mention this.